### PR TITLE
UOE-11092: Support for the imp_counting_method(1px) feature for bidders.

### PR DESCRIPTION
**Pull request by AvinashKapre**

### CODE REVIEW

File: modules/pubmatic/openwrap/publisherfeature/impcountingmethod.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+			}
+		}
+	}
+
+	fe.Lock()
+	fe.impCountingMethod.enabledBidders = enabledBidders
+	fe.Unlock()
+}
+
+func (fe *feature) GetImpCountingMethodEnabledBidders() map[string]struct{} {
```
Comment (Pubmatic-Dhruv-Sonone) [inline modules/pubmatic/openwrap/publisherfeature/impcountingmethod.go:36]: Can we rewrite this as below?
```
func (fe *feature) GetImpCountingMethodEnabledBidders() map[string]struct{} {
    fe.RLock()
    defer fe.RUnlock()
    
    enabledBidders := make(map[string]struct{}, len(fe.impCountingMethod.enabledBidders))
    for bidder := range fe.impCountingMethod.enabledBidders {
        enabledBidders[bidder] = struct{}{}
    }
    return enabledBidders
}
```

File: modules/pubmatic/openwrap/publisherfeature/impcountingmethod_test.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+import (
+	"testing"
+
+	"github.com/prebid/prebid-server/v2/modules/pubmatic/openwrap/cache"
+	"github.com/prebid/prebid-server/v2/modules/pubmatic/openwrap/models"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFeature_updateImpCountingMethodEnabledBidders(t *testing.T) {
```
Comment (Pubmatic-Dhruv-Sonone) [inline modules/pubmatic/openwrap/publisherfeature/impcountingmethod_test.go:11]: Nitpick: `TestFeatureUpdateImpCountingMethodEnabledBidders`

File: modules/pubmatic/openwrap/tracker/inject.go [hunk 3]
```diff
@@ -22,6 +23,7 @@ func InjectTrackers(rctx models.RequestCtx, bidResponse *openrtb2.BidResponse) (
 		for j, bid := range seatBid.Bid {
 			var errMsg string
```
Comment (Pubmatic-Dhruv-Sonone) [inline modules/pubmatic/openwrap/tracker/inject.go:24]: Nitpick: use var block
```
var (
    errMsg string
    err error
    impCountingMethodEnabled bool
)
```

File: modules/pubmatic/openwrap/publisherfeature/impcountingmethod_test.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+				publisherFeature:  tt.fields.publisherFeature,
+				impCountingMethod: tt.fields.impCountingMethod,
+			}
+			fe.updateImpCountingMethodEnabledBidders()
+			assert.Equal(t, tt.wantImpCoutingMethodEnabledBidders, fe.impCountingMethod.enabledBidders)
+		})
+	}
+}
+
+func TestFeature_GetImpCountingMethodEnabledBidders(t *testing.T) {
```
Comment (Pubmatic-Dhruv-Sonone) [inline modules/pubmatic/openwrap/publisherfeature/impcountingmethod_test.go:82]: Nitpick:  `TestFeatureGetImpCountingMethodEnabledBidders`

File: modules/pubmatic/openwrap/publisherfeature/impcountingmethod_test.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+	type fields struct {
+		cache             cache.Cache
+		publisherFeature  map[int]map[int]models.FeatureData
+		impCountingMethod impCountingMethod
+	}
+	tests := []struct {
+		name                               string
+		fields                             fields
+		wantImpCoutingMethodEnabledBidders map[string]struct{}
+	}{
```
Comment (Pubmatic-Dhruv-Sonone) [inline modules/pubmatic/openwrap/publisherfeature/impcountingmethod_test.go:21]: Add ut for following
1. Feature is disabled.
2. Invalid value.

**Review by Pubmatic-Dhruv-Sonone — CHANGES_REQUESTED** [review]

**Review by Pubmatic-Dhruv-Sonone — DISMISSED** [review]

File: modules/pubmatic/openwrap/publisherfeature/reloader.go [hunk 6]
```diff
@@ -117,6 +121,7 @@ func (fe *feature) updateFeatureConfigMaps() {
 	fe.updateAnalyticsThrottling()
 	fe.updateBidRecoveryEnabledPublishers()
 	fe.updateApplovinMultiFloorsFeature()
+	fe.updateImpCountingMethodEnabledBidders()
```
Comment (pm-saurabh-narkhede) [inline modules/pubmatic/openwrap/publisherfeature/reloader.go:124]: Can we have one test case for `updateFeatureConfigMaps`  with `updateImpCountingMethodEnabledBidders` in `Test_feature_updateFeatureConfigMaps`?

**Review by pm-saurabh-narkhede — DISMISSED** [review]

File: modules/pubmatic/openwrap/tracker/inject.go [hunk 7]
```diff
@@ -32,7 +36,9 @@ func InjectTrackers(rctx models.RequestCtx, bidResponse *openrtb2.BidResponse) (
 
 			switch adformat {
 			case models.Banner:
-				bidResponse.SeatBid[i].Bid[j].AdM, bidResponse.SeatBid[i].Bid[j].BURL = injectBannerTracker(rctx, tracker, bid, seatBid.Seat, pixels)
+				if bidResponse.SeatBid[i].Bid[j].AdM, bidResponse.SeatBid[i].Bid[j].BURL, impCountingMethodEnabled = injectBannerTracker(rctx, tracker, bid, seatBid.Seat, pixels); impCountingMethodEnabled {
+					bidResponse.SeatBid[i].Bid[j].Ext, err = jsonparser.Set(bid.Ext, []byte(`1`), models.ImpCountingMethod)
+				}
```
Comment (pm-viral-vala) [inline modules/pubmatic/openwrap/tracker/inject.go:41]: Can we check tracker.IsOMEnabled flag instead of returning boolean flag?

**Review by pm-viral-vala — CHANGES_REQUESTED** [review]

File: modules/pubmatic/openwrap/publisherfeature/impcountingmethod_test.go [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+					},
+				},
+				impCountingMethod: newImpCountingMethod(),
+			},
+			wantImpCoutingMethodEnabledBidders: [2]map[string]struct{}{
+				{},
+				{
+					"appnexus": {},
+					"rubicon":  {},
+				},
```
Comment (pm-viral-vala) [inline modules/pubmatic/openwrap/publisherfeature/impcountingmethod_test.go:55]: Can we have test case if both are already set and after execution of the update function on one index got updated with the latest data? Currently in all test cases expected is initialized with empty data.

File: modules/pubmatic/openwrap/publisherfeature/impcountingmethod_test.go [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+	type fields struct {
+		impCountingMethod impCountingMethod
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		want   map[string]struct{}
+	}{
+		{
+			name: "get imp counting method enabled bidders when index is 0",
```
Comment (pm-viral-vala) [inline modules/pubmatic/openwrap/publisherfeature/impcountingmethod_test.go:170]: Please update all testcases name with underscore instead of spaces.

**Review by pm-viral-vala — CHANGES_REQUESTED** [review]

**Pull request closed without merging**